README.md
# jupyterhub-chef

[![Build Status](https://travis-ci.org/hansohn/jupyterhub-chef.svg?branch=master)](https://travis-ci.org/hansohn/jupyterhub-chef) [![GitHub tag](https://img.shields.io/github/tag/hansohn/jupyterhub-chef.svg)](https://github.com/hansohn/jupyterhub-chef)

This cookbook installs [JupyterHub](https://github.com/jupyterhub/jupyterhub) a multi-user Hub which spawns, manages, and proxies multiple instances of the single-user [Jupyter](http://jupyter.org/) notebook server.

### Prerequsites

By default `Python2` and `Python36` are installed from system package repos. A different Python package, or additional pips, can be specified for installation by overriding the following attributes.

```ruby
# python
node['python']['python2']['packages'] = %w(python-devel python2-pip python-setuptools)
node['python']['python2']['bin'] = '/bin/python2'
node['python']['python2']'pip_bin'] = '/bin/pip2'
node['python']['python2']['pips'] = %w()
node['python']['python3']['packages'] = %w(python3-devel python3-pip python3-setuptools)
node['python']['python3']['bin'] = '/bin/python3.6'
node['python']['python3']'pip_bin'] = '/bin/pip3.6'
node['python']['python3']['pips'] = %w('notebook<6' jupyter py4j ipyparallel)
```

The current `6.x` version of `NodeJS` is installed by default. A different version of NodeJS, or additional npms, can be specified for installation by overriding the following attributes.

```ruby
# node
node['nodejs']['version'] = '6.x'
node['nodejs']['npms'] = %w()
node['nodejs']['global_npms'] = %w(npm configurable-http-proxy)
```

### Configuration

By default this cookbook installs JupyterHub version `0.9.4`, which at the time of this writing, is the current version. Various changes can be made to JupyterHub's configuration by overriding the following attributes.

```ruby
# jupyterhub
node['jupyterhub']['install_from'] = 'python'
node['jupyterhub']['install_version'] = '0.9.6'
node['jupyterhub']['setup']['run_as'] = 'root'
node['jupyterhub']['setup']['pid_file'] = '/var/run/jupyter.pid'
node['jupyterhub']['setup']['app_dir'] = '/opt/jupyterhub'
node['jupyterhub']['setup']['runtime_dir'] = '/srv/jupyterhub'
node['jupyterhub']['setup']['log_dir'] = '/var/log/jupyterhub'
node['jupyterhub']['setup']['enable_ssl'] = false
node['jupyterhub']['setup']['enable_ldap'] = false
node['jupyterhub']['config']['jupyterhub_config'][
node['jupyterhub']['config']['jupyterhub_config']['JupyterHub.ip'] = ''
node['jupyterhub']['config']['jupyterhub_config']['JupyterHub.port'] = '8000'
node['jupyterhub']['config']['jupyterhub_config']['JupyterHub.ssl_port'] = '8443'
node['jupyterhub']['config']['jupyterhub_config']['JupyterHub.hub_ip'] = '127.0.0.1'
node['jupyterhub']['config']['jupyterhub_config']['JupyterHub.hub_port'] = '8081'
node['jupyterhub']['config']['jupyterhub_config']['JupyterHub.proxy_api_ip'] = '127.0.0.1'
node['jupyterhub']['config']['jupyterhub_config']['JupyterHub.proxy_api_port'] = '8001'
node['jupyterhub']['config']['jupyterhub_config']['JupyterHub.ssl_cert'] = '/etc/ssl/certs/jupyterhub.crt'
node['jupyterhub']['config']['jupyterhub_config']['JupyterHub.ssl_key'] = '/etc/ssl/private/jupyterhub.key'
node['jupyterhub']['config']['jupyterhub_config']['Authenticator.whitelist'] = %w()
node['jupyterhub']['config']['jupyterhub_config']['Authenticator.admin_users'] = %w()
node['jupyterhub']['config']['jupyterhub_config']['JupyterHub.authenticator_class'] = 'jupyterhub.auth.PAMAuthenticator'
node['jupyterhub']['config']['jupyterhub_config']['Spawner.cmd'] = 'jupyterhub-singleuser'
node['jupyterhub']['config']['jupyterhub_config']['Spawner.args'] = '--NotebookApp.allow_remote_access=True'
node['jupyterhub']['config']['jupyterhub_config']['Spawner.notebook_dir'] = '~/jupyterhub'
```

### Permissions

All system users with login rights will be allowed to log into JupyterHub. This scope can be narrowed by modifying the following attributes.

```ruby
# permissions
node['jupyterhub']['config']['jupyterhub_config']['Authenticator.whitelist'] = %w()
node['jupyterhub']['config']['jupyterhub_config']['Authenticator.admin_users'] = %w()
```

### LDAP Authentication

LDAP Authentication is not provided in Jupyterhub out of the box. This functionality is provided by the [jupyterhub-ldap-authenticator](https://github.com/hansohn/jupyterhub-ldap-authenticator) package which is installed by default with this cookbook. To utilize this package and enable LDAP Authentication, define the following keys in your attributes file.

```ruby
# enable ldap
node['jupyterhub']['config']['enable_ldap'] = true
node['jupyterhub']['config']['jupyterhub_config']['JupyterHub.authenticator_ldap_class'] = 'ldapauthenticator.LDAPAuthenticator'
node['jupyterhub']['config']['jupyterhub_config']['LDAPAuthenticator.server_hosts'] = %w(ldaps://ldap1.example.com:636 ldaps://ldap2.example.com:636)
node['jupyterhub']['config']['jupyterhub_config']['LDAPAuthenticator.bind_user_dn'] = 'uid=imauser,cn=users,cn=accounts,dc=example,dc=com'
node['jupyterhub']['config']['jupyterhub_config']['LDAPAuthenticator.bind_user_password'] = 'imapassword'
node['jupyterhub']['config']['jupyterhub_config']['LDAPAuthenticator.user_search_base'] = 'cn=users,cn=accounts,dc=example,dc=com'
node['jupyterhub']['config']['jupyterhub_config']['LDAPAuthenticator.user_search_filter'] = '(&(objectClass=person)(uid={username}))'
node['jupyterhub']['config']['jupyterhub_config']['LDAPAuthenticator.user_membership_attribute'] = 'memberOf'
node['jupyterhub']['config']['jupyterhub_config']['LDAPAuthenticator.group_search_base'] = 'cn=groups,cn=accounts,dc=example,dc=com'
node['jupyterhub']['config']['jupyterhub_config']['LDAPAuthenticator.group_search_filter'] = '(&(objectClass=ipausergroup)(memberOf={group}))'
node['jupyterhub']['config']['jupyterhub_config']['LDAPAuthenticator.allowed_groups'] = %w(cn=jupyterhub-users,cn=groups,cn=accounts,dc=example,dc=com)
node['jupyterhub']['config']['jupyterhub_config']['LDAPAuthenticator.allow_nested_groups'] = 'True'
node['jupyterhub']['config']['jupyterhub_config']['LDAPAuthenticator.create_user_home_dir'] = 'True'
node['jupyterhub']['config']['jupyterhub_config']['LDAPAuthenticator.create_user_home_dir_cmd'] = %w(mkhomedir_helper)
```

### PostgreSQL Database

By default Jupyterhub uses a [SQLite](https://www.sqlite.org/index.html) database. To use [PostgreSQL](https://www.postgresql.org/) instead, configure your postgresql server and then add the following keys to your attribute file. Replace values with those associated to your specific instance.

```ruby
node['jupyterhub']['db']['type'] = 'postgresql'
node['jupyterhub']['db']['user'] = 'jupyterhub_db_user'
node['jupyterhub']['db']['pass'] = 'jupyterhub_db_pass'
node['jupyterhub']['db']['host'] = 'jupyterhub_db_server'
node['jupyterhub']['db']['port'] = '5432'
node['jupyterhub']['db']['name'] = 'jupyterhub_db_name'
```

### Kernels

The [IPython](https://ipython.org/) kernel is the Python execution backend for Jupyter/JupyterHub. This cookbook includes 4 kernels by default, they are as follows:

- python2: kernel running native python `2.7`
- python3: kernel running native python `3.6`
- anaconda2: kernel running python `2.7.15` and default anaconda packages
- anaconda3: kernel running python `3.6.5` and default anaconda packages

These kernels can be enabled/disabled as desired. The kernel name, python version, and included pips/condas can also be changed by modifying the following attributes.

```ruby
# python
node['python']['virtualenvs']['python2']['dest_dir'] = '/opt/python/virtualenv/python2'
node['python']['virtualenvs']['python2']['python'] = '/bin/python2'
node['python']['virtualenvs']['python2']['pips'] = %w(boto3 csvkit ipykernel Keras nose nose-parameterized pandas pyGPs requests tensorflow Theano)
node['python']['virtualenvs']['python3']['dest_dir'] = '/opt/python/virtualenv/python3'
node['python']['virtualenvs']['python3']['python'] = '/bin/python3.6'
node['python']['virtualenvs']['python3']['pips'] = %w(boto3 csvkit ipykernel Keras nose nose-parameterized pandas pyGPs requests tensorflow Theano)

# anaconda
node['anaconda']['virtualenvs']['anaconda2']['python'] = '2.7.15'
node['anaconda']['virtualenvs']['anaconda2']['condas'] = %w(numpy pandas)
node['anaconda']['virtualenvs']['anaconda2']['pips'] = %w(ipykernel)
node['anaconda']['virtualenvs']['anaconda3']['python'] = '3.6.5'
node['anaconda']['virtualenvs']['anaconda3']['condas'] = %w(numpy pandas)
node['anaconda']['virtualenvs']['anaconda3']['pips'] = %w(ipykernel)

# jupyter kernels(s)
node['jupyter']['kernels']['python2']['displayname'] = 'Python 2'
node['jupyter']['kernels']['python2']['install'] = true
node['jupyter']['kernels']['python2']['python_dist'] = 'python'
node['jupyter']['kernels']['python2']['python_env'] = 'python2'
node['jupyter']['kernels']['python3']['displayname'] = 'Python 3'
node['jupyter']['kernels']['python3']['install'] = true
node['jupyter']['kernels']['python3']['python_dist'] = 'python'
node['jupyter']['kernels']['python3']['python_env'] = 'python3'
node['jupyter']['kernels']['anaconda2']['displayname'] = 'Anaconda 2'
node['jupyter']['kernels']['anaconda2']['install'] = true
node['jupyter']['kernels']['anaconda2']['python_dist'] = 'anaconda'
node['jupyter']['kernels']['anaconda2']['python_env'] = 'anaconda2'
node['jupyter']['kernels']['anaconda3']['displayname'] = 'Anaconda 3'
node['jupyter']['kernels']['anaconda3']['install'] = true
node['jupyter']['kernels']['anaconda3']['python_dist'] = 'anaconda'
node['jupyter']['kernels']['anaconda3']['python_env'] = 'anaconda3'
```

To create your own custom kernel with a pinned version of python and specific included packages include the following in your attributes file. Replace `python-custom` with the desired name of your kernel.

```ruby
# custom anaconda kernel
node['anaconda']['virtualenvs']['python-custom']['python'] = '3.6.5'
node['anaconda']['virtualenvs']['python-custom']['condas'] = %w(numpy pandas)
node['anaconda']['virtualenvs']['python-custom']['pips'] = %w(ipykernel matplotlib scikit-learn tensorflow)
node['jupyter']['kernels']['python-custom']['displayname'] = 'Python Custom'
node['jupyter']['kernels']['python-custom']['install'] = true
node['jupyter']['kernels']['python-custom']['python_dist'] = 'anaconda'
node['jupyter']['kernels']['python-custom']['python_env'] = 'python-custom'
```

### Usage

Once installed JupyterHub is available at http://127.0.0.1:8000 unless otherwise modified using the attributes referenced above. Before a user can log into JupyterHub he/she must create their notebooks directory at `~/jupyterhub`. Failure to create this directory will result in 500 errors when logging in.

### Documentation

The following resources may be helpful to better understand JupyterHub usage:

- [JupyterHub Website](https://github.com/jupyterhub/jupyterhub)
- [JupyterHub Documentation](https://jupyterhub.readthedocs.io/en/latest/)
- [Jupyter Website](https://jupyter.org/)
- [Jupyter Documentation](https://jupyter.readthedocs.io/en/latest/)
- [Anaconda Website](https://www.continuum.io/)
- [Anaconda Documentation](https://docs.continuum.io/anaconda/)
- [Anaconda Package List](https://docs.continuum.io/anaconda/pkg-docs)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased](unreleased)

- no new features in development at this time

## [3.0.0](https://github.com/hansohn/jupyterhub-chef/compare/2.3.0...3.0.0) (Jan 03, 2020)

FEATURES:

- update jupyterhub to 0.9.6
- update attribute structure
- update anaconda to Anaconda3-2019.10 

## [2.3.0](https://github.com/hansohn/jupyterhub-chef/compare/2.2.0...2.3.0) (Oct 25, 2018)

FEATURES:

- update python to 3.6
- update jupyterhub to 0.9.4

## [2.2.0](https://github.com/hansohn/jupyterhub-chef/compare/2.1.0...2.2.0) (Jul 23, 2018)

FEATURES:

- update license
- bump anaconda version to Anaconda3-5.2.0
- bump anaconda2 kernel to python 2.7.15

BUG FIXES:

- resolve issue where anaconda version attribute is not intrepreted correctly when defined externally

## [2.1.0](https://github.com/hansohn/jupyterhub-chef/compare/2.0.0...2.1.0) (Jun 20, 2018)

FEATURES:

- remove default recipe functionality
- allow any version of jupyterhub to be installed via python or git
- add inspec tests
- add travis ci

## [2.0.0](https://github.com/hansohn/jupyterhub-chef/compare/1.0.0...2.0.0) (Jun 07, 2018)

BREAKING CHANGES:

- implement 'install_from' and 'install_version' attributes to select how to install jupyterhub and which version to install

FEATURES:

- bump default jupyterhub version to 0.8.1
- support jupyterhub installation from python or github. set python as default
- add jupyterhub config parameters for LDAP integration
- add jupyterhub config parameters for PostgreSQL integration
- add Anaconda kernels to jupyterhub by default

BUG FIXES:

- remove 'c.LocalAuthenticator.create_system_users' parameter in config. Doesn't work with our Auth Class

## [1.0.0](https://github.com/hansohn/jupyterhub-chef/compare/0.1.1...1.0.0) (Feb 17, 2017)

BREAKING CHANGES:

- restructure attributes

FEATURES:

- download node directly instead of using nvm
- upgrade default node version to current 6.x

## [0.1.1](https://github.com/hansohn/jupyterhub-chef/compare/0.1.0...0.1.1) (Jun 20, 2016)

BUG FIXES:

- fix initial commit git issue

## [0.1.0](https://github.com/hansohn/jupyterhub-chef/compare/0.1.0...0.1.0) (Jun 20, 2016)

FEATURES:

- initial commit


attributes/default.rb
# node attributes
default['nodejs'] = {
  'version' => '6.x',
  'npms' => %w(),
  'global_npms' => %w(
    npm
    configurable-http-proxy
  ),
}

# python attributes
default['python'] = {
  'python2' => {
    'packages' => %w(
      python-devel
      python2-pip
      python-setuptools
    ),
    'bin' => '/bin/python2',
    'pip_bin' => '/bin/pip2',
    'pips' => %w(),
  },
  'python3' => {
    'packages' => %w(
      python3-devel
      python3-pip
      python3-setuptools
    ),
    'bin' => '/bin/python3.6',
    'pip_bin' => '/bin/pip3.6',
    'pips' => %w(
      'notebook<6'
      jupyter
      py4j
      ipyparallel
    ),
    'symlinks' => {},
  },
  'virtualenvs' => {
    'python2' => {
      'dest_dir' => '/opt/python/virtualenv/python2',
      'python' => '/bin/python2',
      'pips' => %w(
        boto3
        csvkit
        ipykernel
        Keras
        nose
        nose-parameterized
        pandas
        pyGPs
        requests
        tensorflow
        Theano
      ),
    },
    'python3' => {
      'dest_dir' => '/opt/python/virtualenv/python3',
      'python' => '/bin/python3.6',
      'pips' => %w(
        boto3
        csvkit
        ipykernel
        Keras
        nose
        nose-parameterized
        pandas
        pyGPs
        requests
        tensorflow
        Theano
      ),
    },
  },
}

# anaconda attributes
default['anaconda'] = {
  'version' => 'Anaconda3-2019.10',
  'source' => {
    'Anaconda2-5.1.0' => {
      'url' => 'https://repo.continuum.io/archive/Anaconda2-5.1.0-Linux-x86_64.sh',
      'checksum' => '5f26ee92860d1dffdcd20910ff2cf75572c39d2892d365f4e867a611cca2af5b',
    },
    'Anaconda2-5.2.0' => {
      'url' => 'https://repo.continuum.io/archive/Anaconda2-5.2.0-Linux-x86_64.sh',
      'checksum' => 'cb0d7a08b0e2cec4372033d3269979b4e72e2353ffd1444f57cb38bc9621219f',
    },
    'Anaconda3-5.1.0' => {
      'url' => 'https://repo.continuum.io/archive/Anaconda3-5.1.0-Linux-x86_64.sh',
      'checksum' => '7e6785caad25e33930bc03fac4994a434a21bc8401817b7efa28f53619fa9c29',
    },
    'Anaconda3-5.2.0' => {
      'url' => 'https://repo.continuum.io/archive/Anaconda3-5.2.0-Linux-x86_64.sh',
      'checksum' => '09f53738b0cd3bb96f5b1bac488e5528df9906be2480fe61df40e0e0d19e3d48',
    },
    'Anaconda3-2018.12' => {
      'url' => 'https://repo.continuum.io/archive/Anaconda3-2018.12-Linux-x86_64.sh',
      'checksum' => '1019d0857e5865f8a6861eaf15bfe535b87e92b72ce4f531000dc672be7fce00',
    },
    'Anaconda3-2019.10' => {
      'url' => 'https://repo.anaconda.com/archive/Anaconda3-2019.10-Linux-x86_64.sh',
      'checksum' => '46d762284d252e51cd58a8ca6c8adc9da2eadc82c342927b2f66ed011d1d8b53',
    },
  },
  'config' => {
    'app_dir' => '/opt/anaconda',
    'channels' => {
      'add' => %w(conda-forge),
      'remove' => %w(),
    },
  },
  'virtualenvs' => {
    'anaconda2' => {
      'python' => '2.7.15',
      'condas' => %w(
        numpy
        pandas
      ),
      'pips' => %w(
        ipykernel
      ),
    },
    'anaconda3' => {
      'python_version' => '3.6.5',
      'condas' => %w(
        numpy
        pandas
      ),
      'pips' => %w(
        ipykernel
      ),
    },
  },
}

# jupyter attributes
default['jupyter'] = {
  'setup' => {
    'allow_parallel_computing' => true,
    'enable_contrib_nbextensions' => true,
  },
  'kernels' => {
    'python2' => {
      'displayname' => 'Python 2',
      'install' => true,
      'python_dist' => 'python_env',
      'python_env' => 'python2',
    },
    'python3' => {
      'displayname' => 'Python 3',
      'install' => true,
      'python_dist' => 'python_env',
      'python_env' => 'python3',
    },
    'anaconda2' => {
      'displayname' => 'Anaconda 2',
      'install' => true,
      'python_dist' => 'anaconda_env',
      'python_env' => 'anaconda2',
    },
    'anaconda3' => {
      'displayname' => 'Anaconda 3',
      'install' => true,
      'python_dist' => 'anaconda_env',
      'python_env' => 'anaconda3',
    },
  },
}

# jupyterhub attributes
default['jupyterhub'] = {
  'install_from' => 'python',
  'install_version' => '0.9.6',
  'git' => {
    'repo' => 'https://github.com/jupyterhub/jupyterhub',
  },
  'addons' => {
    'packages' => %w(
      blas
      blas-devel
      boost-devel
      gflags-devel
      glog-devel
      hdf5-devel
      leveldb-devel
      lmdb-devel
      opencv-devel
      protobuf-devel
      snappy-devel
    ),
    'pips' => %w(
      ipyparallel
      jupyter_contrib_nbextensions
      jupyter_nbextensions_configurator
      jupyterhub-ldap-authenticator
    ),
    'condas' => %w(),
    'nbextensions' => %w(
      ipyparallel
    ),
    'serverextensions' => %w(
      ipyparallel
    ),
  },
  'group' => {
    'name' => 'jupyterhub',
    'gid' => 10000,
  },
  'user' => {
    'name' => 'jupyterhub',
    'uid' => 15000,
    'home' => '/home/jupyterhub',
    'shell' => '/bin/bash',
  },
  'db' => {
    'type' => 'sqlite',
    'user' => 'jupyterhub_db_user',
    'pass' => 'jupyterhub_db_pass',
    'host' => 'jupyterhub_db_server',
    'port' => '5432',
    'name' => 'jupyterhub_db_name',
  },
  'setup' => {
    'run_as' => 'root',
    'pid_file' => '/var/run/jupyter.pid',
    'app_dir' => '/opt/jupyterhub',
    'runtime_dir' => '/srv/jupyterhub',
    'log_dir' => '/var/log/jupyterhub',
    'enable_ssl' => false,
    'enable_ldap' => false,
  },
  'config' => {
    'jupyterhub_config' => {
      'JupyterHub.ip' => '',
      'JupyterHub.port' => '8000',
      'JupyterHub.ssl_port' => '8443',
      'JupyterHub.hub_ip' => '127.0.0.1',
      'JupyterHub.hub_port' => '8081',
      'JupyterHub.proxy_api_ip' => '127.0.0.1',
      'JupyterHub.proxy_api_port' => '8001',
      'JupyterHub.ssl_cert' => '/etc/ssl/certs/jupyterhub.crt',
      'JupyterHub.ssl_key' => '/etc/ssl/private/jupyterhub.key',
      'Authenticator.whitelist' => %w(),
      'Authenticator.admin_users' => %w(),
      'JupyterHub.authenticator_class' => 'jupyterhub.auth.PAMAuthenticator',
      'JupyterHub.authenticator_ldap_class' => 'ldapauthenticator.LDAPAuthenticator',
      'LDAPAuthenticator.server_hosts' => %w(ldap://ldapserver-1.example.com:389 ldap://ldapserver-2.example.com:389),
      'LDAPAuthenticator.bind_user_dn' => 'uid=ldapquery,cn=users,cn=accounts,dc=example,dc=com',
      'LDAPAuthenticator.bind_user_password' => 'password',
      'LDAPAuthenticator.user_search_base' => 'cn=users,cn=accounts,dc=example,dc=com',
      'LDAPAuthenticator.user_search_filter' => '(&(objectClass=person)(uid={username}))',
      'LDAPAuthenticator.user_membership_attribute' => 'memberOf',
      'LDAPAuthenticator.group_search_base' => 'cn=groups,cn=accounts,dc=example,dc=com',
      'LDAPAuthenticator.group_search_filter' => '(&(objectClass=ipausergroup)(memberOf={group}))',
      'LDAPAuthenticator.allowed_groups' => %w(cn=jupyterhub-users,cn=groups,cn=accounts,dc=example,dc=com),
      'LDAPAuthenticator.allow_nested_groups' => 'True',
      'LDAPAuthenticator.create_user_home_dir' => 'True',
      'LDAPAuthenticator.create_user_home_dir_cmd' => %w(mkhomedir_helper),
      'Spawner.cmd' => 'jupyterhub-singleuser',
      'Spawner.args' => '--NotebookApp.allow_remote_access=True',
      'Spawner.notebook_dir' => '~/jupyterhub',
    },
  },
}
